Clean-up container disk usage output

rvolosatovs opened the pull request:
**- What I did**

- Introduce and use a scoped `types.ContainerUsage` type to avoid returning a lot of unnecessary data on `/system/df` (see example below and integration test changes)
- Refactor and simplify container listing functionality in the daemon to allow ranging over `types.Container` directly without constructing a `[]*types.Container`

**- How to verify it**

```sh
$ docker system df -v
<...>
CONTAINER ID        IMAGE               COMMAND                  LOCAL VOLUMES       SIZE                CREATED ago              STATUS                          NAMES
a6482ad2ce56        alpine              "sh -c 'echo test ..."   1                   5B                  2 seconds ago ago        Exited (0) 1 second ago         xenodochial_antonelli
```


Before:
```
curl -s --unix-socket /var/run/docker.sock http://localhost/system/df?type='container' | jq '.Containers'
[
  {
    "Id": "61578390b6c12b86d1b2c4b327685cfa154e2a6f7802e83e7457d92df33cd072",
    "Names": [
      "/stoic_elbakyan"
    ],
    "Image": "alpine",
    "ImageID": "sha256:021b3423115ff662225e83d7e2606475217de7b55fde83ce3447a54019a77aa2",
    "Command": "sh -c '/bin/echo test > testfile'",
    "Created": 1628607865,
    "Ports": [],
    "SizeRw": 5,
    "SizeRootFs": 5595297,
    "Labels": {},
    "State": "exited",
    "Status": "Exited (0) About a minute ago",
    "HostConfig": {
      "NetworkMode": "default"
    },
    "NetworkSettings": {
      "Networks": {
        "bridge": {
          "IPAMConfig": null,
          "Links": null,
          "Aliases": null,
          "NetworkID": "f124243560fa68100744675eddc958cd0f9cb54451751a5c3cc9855e4261f411",
          "EndpointID": "",
          "Gateway": "",
          "IPAddress": "",
          "IPPrefixLen": 0,
          "IPv6Gateway": "",
          "GlobalIPv6Address": "",
          "GlobalIPv6PrefixLen": 0,
          "MacAddress": "",
          "DriverOpts": null
        }
      }
    },
    "Mounts": [
      {
        "Type": "bind",
        "Source": "/tmp/tmp.35S1m8WikS",
        "Destination": "/some/path/in/container",
        "Mode": "",
        "RW": true,
        "Propagation": "rprivate"
      }
    ]
  }
]
```

After:
```
curl -s --unix-socket /var/run/docker.sock http://localhost/system/df?type='container' | jq '.Containers'
[
  {
    "Id": "61578390b6c12b86d1b2c4b327685cfa154e2a6f7802e83e7457d92df33cd072",
    "Image": "alpine",
    "ImageID": "sha256:021b3423115ff662225e83d7e2606475217de7b55fde83ce3447a54019a77aa2",
    "Command": "sh -c '/bin/echo test > testfile'",
    "Created": 1628607865,
    "Names": [
      "/stoic_elbakyan"
    ],
    "SizeRw": 5,
    "SizeRootFs": 5595297,
    "Status": "Exited (0) 50 seconds ago",
    "Mounts": [
      {
        "Type": "bind",
        "Source": "/tmp/tmp.35S1m8WikS",
        "Destination": "/some/path/in/container",
        "Mode": "",
        "RW": true,
        "Propagation": "rprivate"
      }
    ]
  }
]
```

Commit:
api,daemon: add and use container usage type in DiskUsage

Signed-off-by: Roman Volosatovs <roman.volosatovs@docker.com>

rvolosatovs commented on api/server/router/container/backend.go:
```diff
@@ -55,6 +55,7 @@ type monitorBackend interface {
 	ContainerTop(name string, psArgs string) (*container.ContainerTopOKBody, error)
 
 	Containers(config *types.ContainerListOptions) ([]*types.Container, error)
+	ContainersUsage(ctx context.Context) ([]*types.ContainerUsage, error)
```
Naming is consistent with other methods of this sort (I would prefer `ContainerUsage`)

Commit:
api: preserve DiskUsage.Containers JSON schema pre-1.42

Signed-off-by: Roman Volosatovs <roman.volosatovs@docker.com>

rvolosatovs commented on api/server/router/system/system_routes.go:
```diff
@@ -167,7 +169,26 @@ func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter,
 	}
-	return httputils.WriteJSON(w, http.StatusOK, du)
+	var out interface{} = du
+	if versions.LessThan(version, "1.42") {
```
Should this be 1.42 or 1.43?

rvolosatovs replied:
1.42 is the "next" version, so this is correct

tianon commented:
Isn't this essentially the same as `/containers/json?size=true`?
(although returning _only_ the size data, but I can't imagine the rest of the container metadata is particularly heavy :sweat_smile:)

rvolosatovs commented:
> Isn't this essentially the same as `/containers/json?size=true`?
> (although returning _only_ the size data, but I can't imagine the rest of the container metadata is particularly heavy :sweat_smile:)

At the present time it absolutely is, however @thaJeztah had ideas to put more specific information here, not sure what exactly though

rvolosatovs commented:
> Isn't this essentially the same as `/containers/json?size=true`?
> (although returning _only_ the size data, but I can't imagine the rest of the container metadata is particularly heavy :sweat_smile:)

At the present moment it provides a subset of data returned by `/containers/json?size=true&all=true` or `/system/df?type='container'`, but the intention is to add more data here (not sure what, @thaJeztah ?) and providing an API that is simpler and easier to work with, without the multiple redundant fields (which could potentially also take computational resources to gather, although probably irrelevant compared to size computations)

Commit:
api: ensure empty usage objects are represented as JSON `[]`

Signed-off-by: Roman Volosatovs <roman.volosatovs@docker.com>

rvolosatovs commented on daemon/list.go:
```diff
@@ -174,70 +203,39 @@ func (daemon *Daemon) filterByNameIDMatches(view container.View, ctx *listContex
-		return nil, err
-	}
-
-	// release lock because size calculation is slow
```
Not sure what this is referring to, since there are no locks here; removed the comment

rvolosatovs commented:
CI failure unrealted, caused by Swagger doc gen flakiness:
```
[2021-08-06T15:40:54.567Z] diff --git a/api/types/container/container_wait.go b/api/types/container/container_wait.go

[2021-08-06T15:40:54.567Z] index 49e05ae669..c2e69b5b25 100644

[2021-08-06T15:40:54.567Z] --- a/api/types/container/container_wait.go

[2021-08-06T15:40:54.567Z] +++ b/api/types/container/container_wait.go

[2021-08-06T15:40:54.567Z] @@ -6,14 +6,6 @@ package container // import "github.com/docker/docker/api/types/container"

[2021-08-06T15:40:54.567Z]  // See hack/generate-swagger-api.sh

[2021-08-06T15:40:54.567Z]  // ----------------------------------------------------------------------------

[2021-08-06T15:40:54.567Z]  

[2021-08-06T15:40:54.567Z] -// ContainerWaitOKBodyError container waiting error, if any

[2021-08-06T15:40:54.567Z] -// swagger:model ContainerWaitOKBodyError

[2021-08-06T15:40:54.567Z] -type ContainerWaitOKBodyError struct {

[2021-08-06T15:40:54.567Z] -

[2021-08-06T15:40:54.567Z] -	// Details of an error

[2021-08-06T15:40:54.567Z] -	Message string `json:"Message,omitempty"`

[2021-08-06T15:40:54.567Z] -}

[2021-08-06T15:40:54.567Z] -

[2021-08-06T15:40:54.567Z]  // ContainerWaitOKBody OK response to ContainerWait operation

[2021-08-06T15:40:54.567Z]  // swagger:model ContainerWaitOKBody

[2021-08-06T15:40:54.567Z]  type ContainerWaitOKBody struct {

[2021-08-06T15:40:54.567Z] @@ -26,3 +18,11 @@ type ContainerWaitOKBody struct {

[2021-08-06T15:40:54.567Z]  	// Required: true

[2021-08-06T15:40:54.567Z]  	StatusCode int64 `json:"StatusCode"`

[2021-08-06T15:40:54.567Z]  }

[2021-08-06T15:40:54.567Z] +

[2021-08-06T15:40:54.567Z] +// ContainerWaitOKBodyError container waiting error, if any

[2021-08-06T15:40:54.567Z] +// swagger:model ContainerWaitOKBodyError

[2021-08-06T15:40:54.567Z] +type ContainerWaitOKBodyError struct {

[2021-08-06T15:40:54.567Z] +

[2021-08-06T15:40:54.567Z] +	// Details of an error

[2021-08-06T15:40:54.567Z] +	Message string `json:"Message,omitempty"`

[2021-08-06T15:40:54.567Z] +}

[2021-08-06T15:40:54.567Z] 

[2021-08-06T15:40:54.567Z] Please update api/swagger.yaml with any API changes, then 

[2021-08-06T15:40:54.567Z] run hack/generate-swagger-api.sh.
```

rvolosatovs commented:
Looks like I was too strict when chosing what fields to keep and missed actually used by `docker system df -v`, will update PR

rvolosatovs commented:
Note to reviewers: you probably want `?w=1` for the second commit (link with parameter applied: https://github.com/moby/moby/pull/42711/commits/3194b703a41e02092b401adafb7c46442d7ea611?w=1)
I don't mind also dropping this commit if you think we keep everything as-is and construct a slice first and "trim it down" later, although if we want to do things "right", I think we should not be doing it that way and rather only construct one slice, like is currently suggested in this PR

rvolosatovs commented:
Fixed some commit wording and flipped the order of iteration refactor commit and usage for greater clarity why is (now third) commit necessary https://github.com/moby/moby/pull/42711/commits/cc36e85a1f1e5d2fca006b00a05d184d1e496a70?w=1

thaJeztah commented on api/types/types.go:
```diff
@@ -556,6 +556,21 @@ type DiskUsageOptions struct {
+	SizeRw     int64 `json:",omitempty"`
+	SizeRootFs int64 `json:",omitempty"`
+	Status     string
+	Mounts     []MountPoint
```
We probably can't remove this field if we want it to be backward compatible, but looking at how it's used, it was a bit of a weird choice to return all mount information, because the CLI only does a `count` of volumes that use the `local` driver;

```json
{
      "Mounts": [
        {
          "Type": "volume",
          "Name": "some-volume",
          "Source": "/var/lib/docker/volumes/some-volume/_data",
          "Destination": "/foo",
          "Driver": "local",
          "Mode": "z",
          "RW": true,
          "Propagation": ""
        }
      ]
}
```

CLI code does this; https://github.com/docker/cli/blob/a32cd16160f1b41c1c4ae7bee4dac929d1484e59/cli/command/formatter/container.go#L270-L280

```go
// LocalVolumes returns the number of volumes using the "local" volume driver.
func (c *ContainerContext) LocalVolumes() string {
	count := 0
	for _, m := range c.c.Mounts {
		if m.Driver == "local" {
			count++
		}
	}

	return fmt.Sprintf("%d", count)
}
```

Thinking if it would be worth deprecating the field for API >= 1.42, and instead adding a `LocalVolumes` field (or `VolumeCount`) and do the above logic on the daemon side 🤔 🤔 🤔 🤔 
If we want to do that, we could make it a pointer, and `omitempty`, and clear the field on < 1.42

```suggestion
	Mounts     *[]MountPoint  `json:",omitempty"`
```

rvolosatovs replied:
Sounds like a good idea, but let's leave that for a separate PR and focus just on "trimming down" here without changing the existing functionality too much

thaJeztah commented on api/types/types.go:
```diff
@@ -556,6 +556,21 @@ type DiskUsageOptions struct {
+	Image      string
+	ImageID    string
+	Command    string
+	Created    int64
```
Looks like we already did this for regular containers, so probably should be kept, but `time.Time` would've been good for this (as for `Volumes.CreatedAt`).

We could consider adding `json:",omitempty"` to these fields though

rvolosatovs replied:
Let's leave that for a separate PR and focus just on "trimming down" here without changing the existing functionality too much

thaJeztah commented on api/types/types.go:
```diff
@@ -556,6 +556,21 @@ type DiskUsageOptions struct {
+	Names      []string
+	SizeRw     int64 `json:",omitempty"`
+	SizeRootFs int64 `json:",omitempty"`
+	Status     string
```
(Just a random comment) the `Status` field is kinda horrible, as it's pre-formatting the output (e.g. `"Up 4 seconds"`). We don't have a good solution for that yet though (perhaps we should add a `ServerTime` header to API responses, so that the `container.State` field can be returned, and relative times can be handled client side. I know there's some bugs in that area in the CLI currently (wrong duration shown because it compares "local" time with "server-based" times)

thaJeztah commented on daemon/disk_usage.go:
```diff
@@ -34,6 +29,10 @@ func (daemon *Daemon) ContainerDiskUsage(ctx context.Context) ([]*types.Containe
+			lCtx.idx++
+			return true
+		}); err != nil {
+			return nil, fmt.Errorf("failed to range over containers: %v", err)
```
I need to take some time looking at this commit to get the full picture.

Just a quick comment on this one; wondering if this is more descriptive than the old error.

One thing I _did_ notice is that we're masking the underlying error type if the filter is invalid; `reduceContainers` / `rangeContainers` calls `config.Filters.Validate`, which can return an `errdefs.InvalidParameter`. Using `fmt.Errorf` here will change a `4xx` status to a `5xx` status if that happens.

rvolosatovs commented on api/swagger.yaml:
```diff
@@ -4090,6 +4090,27 @@ definitions:
           items:
             $ref: "#/definitions/Mount"
 
+  ContainerUsage:
+    type: "object"
```
This was just copy-pasted from `Container` with some fields were removed, will update!

rvolosatovs commented on api/swagger.yaml:
```diff
@@ -4090,6 +4090,27 @@ definitions:
+      Id:
+        description: "The ID of this container"
+        type: "string"
+        x-go-name: "ID"
```
Should we do the same in the original `Container` then?

rvolosatovs replied:
Let's leave the `Container` part for a separate PR

rvolosatovs commented on api/server/router/system/system_routes.go:
```diff
@@ -170,7 +170,32 @@ func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter,
+	if versions.LessThan(version, "1.42") {
+		containers := make([]types.Container, 0, len(du.Containers))
+		for _, c := range du.Containers {
+			containers = append(containers, types.Container{
+				ID:         c.ID,
+				Image:      c.Image,
```
Yes, see https://github.com/moby/moby/pull/42711#discussion_r683629561

Commit:
api: move DiskUsage examples to respective type sections

Signed-off-by: Roman Volosatovs <roman.volosatovs@docker.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

Commit:
api: extract ContainerStateStatus

Signed-off-by: Roman Volosatovs <roman.volosatovs@docker.com>

rvolosatovs commented:
Added some missing fields to account for `docker system df --format=raw` usages

I also extracted `ContainerStateStatus` (https://github.com/moby/moby/pull/42711/commits/97b53e4c310fd50cca5f16698af7d2f133b062b1) to reuse the definition with enums values etc.

I think that the "correct" approach would be to extract all of these fields in a similar way and reuse with e.g. `ContainerSummary` to avoid duplication and ensure consistency of (Go) typing, description, etc. , but @thaJeztah was against that in an offline conversation, so not doing that here.

rvolosatovs commented:
Everything passes except for `docker-py` tests:
```
[2021-08-12T19:16:11.523Z] =================================== FAILURES ===================================

[2021-08-12T19:16:11.523Z] _________________ ServiceTest.test_create_service_with_config __________________

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:722: in test_create_service_with_config

[2021-08-12T19:16:11.523Z]     container = self.get_service_container(name)

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:46: in get_service_container

[2021-08-12T19:16:11.523Z]     all=include_stopped

[2021-08-12T19:16:11.523Z] docker/api/container.py:214: in containers

[2021-08-12T19:16:11.523Z]     return [{'Id': x['Id']} for x in res]

[2021-08-12T19:16:11.523Z] E   TypeError: 'NoneType' object is not iterable

[2021-08-12T19:16:11.523Z] _________________ ServiceTest.test_create_service_with_secret __________________

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:670: in test_create_service_with_secret

[2021-08-12T19:16:11.523Z]     container = self.get_service_container(name)

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:46: in get_service_container

[2021-08-12T19:16:11.523Z]     all=include_stopped

[2021-08-12T19:16:11.523Z] docker/api/container.py:214: in containers

[2021-08-12T19:16:11.523Z]     return [{'Id': x['Id']} for x in res]

[2021-08-12T19:16:11.523Z] E   TypeError: 'NoneType' object is not iterable

[2021-08-12T19:16:11.523Z] _____________ ServiceTest.test_create_service_with_unicode_config ______________

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:747: in test_create_service_with_unicode_config

[2021-08-12T19:16:11.523Z]     container = self.get_service_container(name)

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:46: in get_service_container

[2021-08-12T19:16:11.523Z]     all=include_stopped

[2021-08-12T19:16:11.523Z] docker/api/container.py:214: in containers

[2021-08-12T19:16:11.523Z]     return [{'Id': x['Id']} for x in res]

[2021-08-12T19:16:11.523Z] E   TypeError: 'NoneType' object is not iterable

[2021-08-12T19:16:11.523Z] _____________ ServiceTest.test_create_service_with_unicode_secret ______________

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:695: in test_create_service_with_unicode_secret

[2021-08-12T19:16:11.523Z]     container = self.get_service_container(name)

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:46: in get_service_container

[2021-08-12T19:16:11.523Z]     all=include_stopped

[2021-08-12T19:16:11.523Z] docker/api/container.py:214: in containers

[2021-08-12T19:16:11.523Z]     return [{'Id': x['Id']} for x in res]

[2021-08-12T19:16:11.523Z] E   TypeError: 'NoneType' object is not iterable

[2021-08-12T19:16:11.523Z] ________________________ ServiceTest.test_service_logs _________________________

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:136: in test_service_logs

[2021-08-12T19:16:11.523Z]     assert self.get_service_container(name, include_stopped=True)

[2021-08-12T19:16:11.523Z] tests/integration/api_service_test.py:46: in get_service_container

[2021-08-12T19:16:11.523Z]     all=include_stopped

[2021-08-12T19:16:11.523Z] docker/api/container.py:214: in containers

[2021-08-12T19:16:11.523Z]     return [{'Id': x['Id']} for x in res]

[2021-08-12T19:16:11.523Z] E   TypeError: 'NoneType' object is not iterable

[2021-08-12T19:16:11.523Z] ------- generated xml file: /src/bundles/test-docker-py/junit-report.xml -------

[2021-08-12T19:16:11.523Z] =========================== short test summary info ============================

[2021-08-12T19:16:11.523Z] XFAIL tests/integration/api_container_test.py::CreateContainerTest::test_create_with_cpu_rt_options

[2021-08-12T19:16:11.523Z]   CONFIG_RT_GROUP_SCHED isn't enabled

[2021-08-12T19:16:11.523Z] XFAIL tests/integration/api_container_test.py::CreateContainerTest::test_create_with_storage_opt

[2021-08-12T19:16:11.523Z]   Not supported on most drivers

[2021-08-12T19:16:11.523Z] XFAIL tests/integration/api_container_test.py::ContainerTopTest::test_top

[2021-08-12T19:16:11.523Z]   Output of docker top depends on host distro, and is not formalized.

[2021-08-12T19:16:11.523Z] XFAIL tests/integration/api_container_test.py::ContainerTopTest::test_top_with_psargs

[2021-08-12T19:16:11.523Z]   Output of docker top depends on host distro, and is not formalized.

[2021-08-12T19:16:11.523Z] XFAIL tests/integration/api_swarm_test.py::SwarmTest::test_init_swarm_data_path_addr

[2021-08-12T19:16:11.523Z]   Can fail if eth0 has multiple IP addresses

[2021-08-12T19:16:11.523Z] XFAIL tests/integration/api_swarm_test.py::SwarmTest::test_init_swarm_with_log_driver

[2021-08-12T19:16:11.523Z]   This doesn't seem to be taken into account by the engine

[2021-08-12T19:16:11.523Z] SKIPPED [1] tests/integration/api_image_test.py:291: Doesn't work inside a container - FIXME

[2021-08-12T19:16:11.523Z] SKIPPED [1] tests/integration/api_service_test.py:473: API version is too low (< 1.40)

[2021-08-12T19:16:11.523Z] SKIPPED [1] /src/tests/integration/api_swarm_test.py:31: Test stalls the engine on 1.12.0

[2021-08-12T19:16:11.523Z] = 5 failed, 373 passed, 3 skipped, 1 deselected, 6 xfailed, 2 xpassed in 327.15 seconds =

[2021-08-12T20:56:10.124Z] Sending interrupt signal to process

[2021-08-12T20:56:20.963Z] Terminated

[2021-08-12T20:56:20.968Z] script returned exit code 143
```

Not sure what's happening here. It looks like the python SDK is calling `/containers/json`, not `/system/df`, e.g. https://github.com/docker/docker-py/blob/1abeb46dfa1dbe89e384652ef81e57def0b2e285/docker/api/container.py#L208-L216

I will investigate further later. Meanwhile, maybe anyone has an idea what's happening here?

rvolosatovs commented on integration/system/disk_usage_test.go:
```diff
@@ -89,22 +93,17 @@ func TestDiskUsage(t *testing.T) {
-							HostConfig:      du.Containers[0].HostConfig,      // not relevant for the test
-							NetworkSettings: du.Containers[0].NetworkSettings, // not relevant for the test
-							Mounts:          du.Containers[0].Mounts,          // not relevant for the test
+							ID:         cID,
```
I will still add the missing fields here

rvolosatovs commented on docs/api/version-history.md:
```diff
@@ -17,6 +17,10 @@ keywords: "API, Docker, rcli, REST, documentation"
 
 [Docker Engine API v1.42](https://docs.docker.com/engine/api/v1.42/) documentation
 
+* Removed `Labels`, `HostConfig`, `NetworkSettings`, `Ports` and `State` fields
```
Will update this

rvolosatovs commented:
CI failed with (so far):
```
[2021-08-17T17:44:39.926Z] #57 [containerd 2/2] RUN --mount=type=cache,target=/root/.cache/go-build     --mount=type=cache,target=/go/pkg/mod     --mount=type=bind,src=hack/dockerfile/install,target=/tmp/install         PREFIX=/build /tmp/install/install.sh containerd

[2021-08-17T17:44:39.926Z] #57 sha256:a10fe367e8740df5bc8009296ed8b3e6f31b51e18644d8e7b0399c14c594c720

[2021-08-17T17:44:39.926Z] #57 26.24 error: RPC failed; curl 56 GnuTLS recv error (-54): Error in the pull function.

[2021-08-17T17:44:39.926Z] #57 26.24 fatal: the remote end hung up unexpectedly

[2021-08-17T17:44:39.926Z] #57 26.28 fatal: early EOF

[2021-08-17T17:44:39.926Z] #57 26.30 fatal: index-pack failed

[2021-08-17T17:44:39.926Z] #57 ERROR: executor failed running [/bin/sh -c PREFIX=/build /tmp/install/install.sh containerd]: exit code: 128
```

thaJeztah reviewed (commented):
Thanks; I left a couple of comments; pushed some commits to https://github.com/moby/moby/compare/master...thaJeztah:container_usage_TEMP_WITH_SUGGESTIONS?expand=1

Still a bit concerned about the amount of changes in the refactor, and on the fence if it outweighs the benefits.

I was also thinking about a very naive approach to deprecate these fields in the API response, but haven't given that much thought yet; https://github.com/moby/moby/compare/master...thaJeztah:container_usage_simple_approach?expand=1

Perhaps the first couple of commits (swagger changes) could make sense as a separate PR?

thaJeztah reviewed (changes requested):
Found one issue in the third commit ("api: add container usage type"); I think it would also make sense to squash the third and fourth commit ("api,daemon: use container usage type in DiskUsage"), as they're both introducing the new type (one in the swagger, the other the actual code).

Otherwise LGTM

thaJeztah reviewed (approved):
LGTM

did a quick rebase, and squashed the last fix-up commit with the commit that added it
